demo/api/src/products/product-persist.pipe.ts

```diff
@@ -0,0 +1,129 @@
+            //     onConflictExcludeFields: ["id", "createdAt"],
+            // });
+            if (record) {
+                await this.logger.log(`Existing record for ${JSON.stringify(updateQuery)} with id ${record.id} found, updating...`);
+                // keeping the id from the input data (in the DataTransformer plainToInstance always creates a new Entity instance, which is populated with the new id)
+                data.id = record.id;
+                await this.em.nativeUpdate(Product.name, { id: record.id }, { ...productData, category, id: data.id, updatedAt: new Date() });
+            } else {
+                // in case the id is not provided
+                data.id = data?.id || v4();
+                await this.logger.log(
+                    `${updateQuery ? `No record for query ${JSON.stringify(updateQuery)} found` : "No query provided"}, inserting with id ${
+                        data.id
+                    }...`,
+                );
+                await this.em.insert(Product, {
+                    ...productData,
+                    id: data.id,
+                    category,
+                    createdAt: new Date(),
+                    updatedAt: new Date(),
+                });
+                await this.logger.log(`Inserted record with id ${data.id}`, false);
+            }
+
+            let colorsArray: { hex: string; name: string }[] = [];
+            try {
+                colorsArray = JSON.parse(colorString);
+            } catch (error) {
+                await this.logger.warn(`Invalid JSON string for colors: "${colorString}". Defaulting to an empty array.`);
+            }
+
+            await this.em.nativeDelete(ProductColor, { product: data.id });
+
+            if (colorsArray.length > 0) {
+                const productColors = colorsArray.map((color) => ({
+                    id: v4(),
+                    name: color.name,
+                    hexCode: color.hex,
+                    product: data.id,
+                    createdAt: new Date(),
+                    updatedAt: new Date(),
+                }));
+
+                await this.em.insertMany(ProductColor, productColors);
+            }
+
+            this.persistedEntitiesAmount++;
+            await this.logger.log(`Persisted entities amount: ${this.persistedEntitiesAmount}`);
+            // clear entity manager to prevent memory leak
+            if (this.persistedEntitiesAmount % 100 === 0) {
+                this.em.clear();
+            }
+            const outputData = inputDataAndMetadata;
+            outputData.metadata.persistedEntitiesAmount = this.persistedEntitiesAmount;
+            this.push(outputData);
+        } catch (error: unknown) {
+            await this.logger.error(`Error persisting Data: ${error}`);
+            if (error instanceof Error) {
+                callback(error);
+            } else {
+                callback(new Error(`An unknown error occurred: ${error}`));
+            }
```

Comment:
We probably need a return here, otherwise callback is called twice.

Reply:
I've tried it. We don't need a return. But i won't hurt to have them.